sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -518,7 +518,25 @@ object RemoveNoopOperators extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan.transformUpWithPruning(
     _.containsAnyPattern(PROJECT, WINDOW), ruleId) {
     // Eliminate no-op Projects
-    case p @ Project(_, child) if child.sameOutput(p) => child
+    case p @ Project(output, child) if child.sameOutput(p) =>
+      child match {
+        case relation: LeafNode
+          if !output.zip(relation.output).forall { case (a1, a2) => a1.name == a2.name } =>
+          p
+        case _ =>
+          val ret = child.transformExpressionsDown {
+            case named: NamedExpression =>
```

Comment:
I think we need to be explicit here. We can only change the output name if `child` has a project list, which is `Project` and `Aggregate`. We don't need to handle the general `NamedExpression`, as the project list can only be `Attribute` or `Alias`.

Reply:
> I think we need to be explicit here. We can only change the output name if `child` has a project list, which is `Project` and `Aggregate`. We don't need to handle the general `NamedExpression`, as the project list can only be `Attribute` or `Alias`.

Yea, You have explained a point that I am puzzled a lot. How about current.

Reply:
Changed